packages/kbn-data-stream-adapter/src/create_or_update_component_template.ts

```diff
@@ -0,0 +1,112 @@
+  ClusterPutComponentTemplateRequest,
+  IndicesGetIndexTemplateIndexTemplateItem,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { Logger, ElasticsearchClient } from '@kbn/core/server';
+import { asyncForEach } from '@kbn/std';
+import { retryTransientEsErrors } from './retry_transient_es_errors';
+
+interface CreateOrUpdateComponentTemplateOpts {
+  logger: Logger;
+  esClient: ElasticsearchClient;
+  template: ClusterPutComponentTemplateRequest;
+  totalFieldsLimit: number;
+}
+
+const putIndexTemplateTotalFieldsLimitUsingComponentTemplate = async (
+  esClient: ElasticsearchClient,
+  componentTemplateName: string,
+  totalFieldsLimit: number,
+  logger: Logger
+) => {
+  // Get all index templates and filter down to just the ones referencing this component template
+  const { index_templates: indexTemplates } = await retryTransientEsErrors(
+    () => esClient.indices.getIndexTemplate(),
+    { logger }
+  );
+  const indexTemplatesUsingComponentTemplate = (indexTemplates ?? []).filter(
+    (indexTemplate: IndicesGetIndexTemplateIndexTemplateItem) =>
+      (indexTemplate.index_template?.composed_of ?? []).includes(componentTemplateName)
+  );
+
+  await asyncForEach(
+    indexTemplatesUsingComponentTemplate,
+    async (template: IndicesGetIndexTemplateIndexTemplateItem) => {
+      await retryTransientEsErrors(
+        () =>
+          esClient.indices.putIndexTemplate({
+            name: template.name,
+            body: {
+              ...template.index_template,
+              template: {
+                ...template.index_template.template,
+                settings: {
+                  ...template.index_template.template?.settings,
+                  'index.mapping.total_fields.limit': totalFieldsLimit,
+                },
+              },
+            },
+          }),
+        { logger }
+      );
+    }
+  );
+};
+
+const createOrUpdateComponentTemplateHelper = async (
+  esClient: ElasticsearchClient,
+  template: ClusterPutComponentTemplateRequest,
+  totalFieldsLimit: number,
+  logger: Logger
+) => {
+  try {
+    await retryTransientEsErrors(() => esClient.cluster.putComponentTemplate(template), { logger });
+  } catch (error) {
+    if (error.message.match(/Limit of total fields \[\d+\] has been exceeded/) != null) {
+      // This error message occurs when there is an index template using this component template
+      // that contains a field limit setting that using this component template exceeds
+      // Specifically, this can happen for the ECS component template when we add new fields
+      // to adhere to the ECS spec. Individual index templates specify field limits so if the
+      // number of new ECS fields pushes the composed mapping above the limit, this error will
+      // occur. We have to update the field limit inside the index template now otherwise we
+      // can never update the component template
+      await putIndexTemplateTotalFieldsLimitUsingComponentTemplate(
```

Comment:
Consider adding info logging that indicates we're updating the total fields limit for `template.name` from the previous limit (captured from the regex `match`) to the new limit specified by `totalFieldsLimit`